src/vmm/src/devices/virtio/net/metrics.rs

```diff
@@ -0,0 +1,349 @@
+
+use std::collections::BTreeMap;
+use std::sync::{RwLock, RwLockReadGuard};
+
+use serde::ser::SerializeMap;
+use serde::{Serialize, Serializer};
+
+use crate::logger::{IncMetric, SharedIncMetric};
+
+/// returns lock-unwrapped NET_DEV_METRICS_PVT.metrics,
+/// we have this function so that we don't have to make
+/// NET_DEV_METRICS_PVT public
+pub fn get_net_metrics() -> RwLockReadGuard<'static, NetMetricsPerDevice> {
+    // lock is always initialized so it is safe the unwrap the lock without a check.
+    NET_DEV_METRICS_PVT.read().unwrap()
+}
+
+/// map of network interface id and metrics
+/// this should be protected by a lock before accessing.
+#[derive(Debug)]
+pub struct NetMetricsPerDevice {
+    /// used to access per net device metrics
+    pub metrics: BTreeMap<String, NetDeviceMetrics>,
+}
+
+impl NetMetricsPerDevice {
+    /// Allocate `NetDeviceMetrics` for net device having
+    /// id `iface_id`. Also, allocate only if it doesn't
+    /// exist to avoid overwriting previously allocated data.
+    /// lock is always initialized so it is safe the unwrap
+    /// the lock without a check.
+    pub fn alloc(iface_id: String) {
+        if NET_DEV_METRICS_PVT
+            .read()
+            .unwrap()
+            .metrics
+            .get(&iface_id)
+            .is_none()
+        {
+            NET_DEV_METRICS_PVT
+                .write()
+                .unwrap()
+                .metrics
+                .insert(iface_id, NetDeviceMetrics::default());
+        }
+    }
+}
+
+#[macro_export]
+// Per device metrics are behind a ref lock and in a map which
+// makes it difficult to return the metrics pointer via a function
+// This macro is an attempt to make the code more readable and to
+// centralize the error handling.
+// The macro can be called in below ways where iface_id is String and
+// activate_fails is a metric from struct `NetDeviceMetrics`:
+//     net_metrics!(&iface_id, activate_fails.inc());
+//     net_metrics!(&iface_id, activate_fails.add(5));
+//     net_metrics!(&iface_id, activate_fails.count());
+macro_rules! net_metrics {
+    ($iface_id:expr,$metric:ident.$action:ident($($value:tt)?)) => {{
+        if $crate::devices::virtio::net::metrics::get_net_metrics().metrics.get($iface_id).is_some() {
+            $crate::devices::virtio::net::metrics::get_net_metrics().metrics.get($iface_id).unwrap().$metric.$action($($value)?)
+        }else{
+            $crate::devices::virtio::net::metrics::NetMetricsPerDevice::alloc($iface_id.to_string());
+            $crate::devices::virtio::net::metrics::get_net_metrics().metrics.get($iface_id).unwrap().$metric.$action($($value)?)
+        }
+    }}
+}
+
+/// Pool of Network-related metrics per device behind a lock to
+/// keep things thread safe. Since the lock is initialized here
+/// it is safe to unwrap it without any check.
+static NET_DEV_METRICS_PVT: RwLock<NetMetricsPerDevice> = RwLock::new(NetMetricsPerDevice {
+    metrics: BTreeMap::new(),
+});
```

Comment:
nit: maybe rename `NetMetricsPerDevice` -> `NetMetrics`, `NET_DEV_METRICS_PVT` -> `GLOBAL_NET_METRICS` (because I have no idea what PVT means).

Reply:
PVT stands for private :)

Reply:
typically _g or _global is added for global variables and _pvt or _private is added for private variables. Its used extensively in linux and c where adding visibility as suffix helps make the code more readable.
But assuming this is not something we do in rust I can remove the _pvt suffix.

Reply:
But I prefer NetMetricsPerDevice to NetMetrics since it better describes that it is meant for per device metrics so it would have a map/vec in it.